### PR TITLE
Add IPM support for stm32mp157

### PR DESCRIPTION
**Pull request by arnopo**
Add mailbox management based on the IPCC Peripheral present in the stm32MP1 SoC.

Related PR: https://github.com/zephyrproject-rtos/hal_stm32/pull/1

### CODE REVIEW

**Review by Mani-Sadhasivam — CHANGES_REQUESTED** [review]
Please split soc and board support files into separate commits.

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (Mani-Sadhasivam) [inline drivers/ipm/ipm_stm32_ipcc.c]: TX interrupt could be enabled in `stm32_ipcc_mailbox_ipm_send`? I don't see a reason to enable it here.

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (Mani-Sadhasivam) [reply drivers/ipm/ipm_stm32_ipcc.c]: Also, not clearing RX status flag here?

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (Mani-Sadhasivam) [inline drivers/ipm/ipm_stm32_ipcc.c]: Buffer is unused? So no data is being sent?

**Review by Mani-Sadhasivam — CHANGES_REQUESTED** [review]

File: drivers/ipm/Kconfig
Comment (ulfalizer) [inline drivers/ipm/Kconfig]: Redundant `default n`. See [this page](https://docs.zephyrproject.org/latest/guides/kconfig/index.html#redundant-defaults).

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (erwango) [inline drivers/ipm/ipm_stm32_ipcc.c]: Since data->clock is not used anywhere else, can't we call directly `device_get_binding(STM32_CLOCK_CONTROL_NAME);` here, spare clock field and _stm32_ipcc_get_clock function?

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (erwango) [inline drivers/ipm/ipm_stm32_ipcc.c]: Use of `IPCC` is correct, but hidden. I'd prefer we stick with use of dt extracted DT_INST_0_ST_STM32_IPCC_MAILBOX_BASE_ADDRESS from a `config->base` field.
And then spare the data->hipcc (if possible).

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (erwango) [inline drivers/ipm/ipm_stm32_ipcc.c]: 'Inform remote'

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (erwango) [inline drivers/ipm/ipm_stm32_ipcc.c]: Maybe we can also check size, there's a dedicated error value: `EMSGSIZE`

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (erwango) [inline drivers/ipm/ipm_stm32_ipcc.c]: Comment doesn't match the code.
Also, we might want to implement the EBUSY API behavior

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (arnop2) [reply drivers/ipm/ipm_stm32_ipcc.c]: yes you right copy/past issue ...
Concerning the RX status, do you see any reason to clean it on disable?  I would not lost an IRQ on a temporary disable...

File: drivers/ipm/ipm_stm32_ipcc.c
Comment (erwango) [inline drivers/ipm/ipm_stm32_ipcc.c]: You might want to remove/rewrite this line.

File: boards/arm/stm32mp157c_dk2/stm32mp157c_dk2_defconfig
Comment (erwango) [inline boards/arm/stm32mp157c_dk2/stm32mp157c_dk2_defconfig]: Do we consider it should be enabled by default?

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/ipm/st,stm32-ipcc-mailbox.yaml
Comment (erwango) [inline dts/bindings/ipm/st,stm32-ipcc-mailbox.yaml]: Please remove this line

Comment (erwango) [inline dts/bindings/ipm/st,stm32-ipcc-mailbox.yaml]: Please add your Copyright

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/ipm/st,stm32-ipcc-mailbox.yaml
Comment (erwango) [inline dts/bindings/ipm/st,stm32-ipcc-mailbox.yaml]: Please remove this.
cf https://github.com/zephyrproject-rtos/zephyr/pull/16913

**Review by erwango — CHANGES_REQUESTED** [review]
One last comment.
Otherwise LGTM

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
Approved, but DNM until hal_stm32 is merged and PR updated with matching SHA1

**Comment by erwango** [comment]
@Mani-Sadhasivam , can you review again?

**Review by Mani-Sadhasivam — APPROVED** [review]
LGTM

**Comment by arnop2** [comment]
west.yml updated with  SHA1 dependency for hal_stm32  module.

**Review by erwango — CHANGES_REQUESTED** [review]
Let's merge https://github.com/zephyrproject-rtos/zephyr/pull/16984 first.
Please remove hal_stm32 version update commit to avoid conflicts.

File: drivers/ipm/Kconfig
Comment (erwango) [inline drivers/ipm/Kconfig]: Can you break a line here?

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]

File: boards/arm/stm32mp157c_dk2/Kconfig.defconfig
Comment (erwango) [inline boards/arm/stm32mp157c_dk2/Kconfig.defconfig]: This should be moved in SoC as I don't think another MP1 board will support other config

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/stm32mp157c_dk2/Kconfig.defconfig
Comment (arnop2) [reply boards/arm/stm32mp157c_dk2/Kconfig.defconfig]: done

**Review by erwango — APPROVED** [review]

**Comment by arnop2** [comment]
@galak,  @nashif. Please could you review and merge if is ok for you?
Thanks in advance

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
LGTM upon light inspection, merging as STM32 maintainer @erwango has approved it and mentions is required for further progress

**Pull request merged**
